### PR TITLE
Sampler animates while running (PT-186764176)

### PR DESCRIPTION
**Pull request by lublagg**
This PR adds a hook and state property to manage animations during a run.

So far, the aspects of animation addressed are:
- The CODAP table opens when a run is first begun
- The speed determines the length of each step of the animation
- The speed can be adjusted during a run and this affects the look of the mixer animation
- The mixer animates during a run in the same way as the old Sampler
- After each sample is collected, results are sent to the CODAP table
- Controls are disabled while animation is running

Some remaining tasks to be done:
- If user sets speed to "fastest" during run, all remaining item need to be sent to CODAP
- Animating spinner during a run
- Animating variables moving into position below the column name
- Animating variables moving into brackets (also create the brackets component)
- Animating arrows in order

### CODE REVIEW

File: src/components/model/device-views/shared/balls.tsx [hunk 1]
```diff
@@ -11,50 +13,267 @@ interface IBalls {
...
+      let timeoutId: number;
+      let scrambledInitialSetup = false;
+      let skipOffset = stepOffset % ballsArray.length;
+
+
+      const animateMixer = () => {
+        const timeElapsed = Date.now() - startTime;
+        if (deviceAnimationStep?.id === deviceId) {
+          stepOffset += 1;
+          if (deviceAnimationStep?.id === deviceId) {
```
Comment (dougmartin) [inline src/components/model/device-views/shared/balls.tsx:80]: This seems to be a repeat of line 78 (two lines above)

File: src/components/model/device-views/shared/balls.tsx [hunk 2]
```diff
@@ -11,50 +13,267 @@ interface IBalls {
...
+            timeoutId = setTimeout(() => {
+              animationFrameId = requestAnimationFrame(animateMixer);
+            }, timeout);
+          }
+
+          if (deviceAnimationStep?.id === deviceId) {
```
Comment (dougmartin) [inline src/components/model/device-views/shared/balls.tsx:86]: Also a repeat of line 78?

**Review by dougmartin — COMMENTED** [review]
Looks good so far! :+1:

**Review by eireland — APPROVED** [review]
Wow! Great work!
I'm approving this one and you can do a separate PR for the balls going out of bounds during animation.

**Pull request merged**